packages/graph/hash_graph/lib/graph/src/types/schema/data.rs

```diff
@@ -0,0 +1,316 @@
+use std::collections::HashMap;
+
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+
+use crate::types::schema::{Uri, Validate, ValidationError};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(
+    try_from = "DataTypeRepr",
+    tag = "kind",
+    rename = "dataType",
+    rename_all = "camelCase"
+)]
+pub struct DataType {
+    #[serde(rename = "$id")]
+    id: Uri,
+    title: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    description: Option<String>,
+    #[serde(rename = "type")]
+    ty: String,
+    #[serde(flatten)]
+    additional_properties: HashMap<String, serde_json::Value>,
+}
+
+impl DataType {
+    /// Creates a new `PropertyType` without validating.
+    #[must_use]
+    pub fn new_unchecked(
+        id: Uri,
+        title: impl Into<String>,
+        description: impl Into<Option<String>>,
+        ty: impl Into<String>,
+        additional_properties: HashMap<String, serde_json::Value>,
+    ) -> Self {
+        Self {
+            id,
+            title: title.into(),
+            description: description.into(),
+            ty: ty.into(),
+            additional_properties,
+        }
+    }
+
+    /// Creates a new `PropertyType`.
+    ///
+    /// # Errors
+    ///
+    /// - [`ValidationError`] if validation is failing.
+    pub fn new(
+        id: Uri,
+        title: impl Into<String>,
+        description: impl Into<Option<String>>,
+        ty: impl Into<String>,
+        additional_properties: HashMap<String, serde_json::Value>,
+    ) -> Result<Self, ValidationError> {
+        let property_type = Self::new_unchecked(id, title, description, ty, additional_properties);
+        property_type.validate()?;
+        Ok(property_type)
+    }
+
+    #[must_use]
+    pub const fn id(&self) -> &Uri {
+        &self.id
+    }
+
+    #[must_use]
+    pub fn title(&self) -> &str {
+        &self.title
+    }
+
+    #[must_use]
+    pub fn description(&self) -> Option<&str> {
+        self.description.as_deref()
+    }
+
+    #[must_use]
+    pub fn ty(&self) -> &str {
+        &self.ty
+    }
+
+    #[must_use]
+    pub const fn additional_properties(&self) -> &HashMap<String, serde_json::Value> {
+        &self.additional_properties
+    }
+
+    #[must_use]
+    pub fn additional_properties_mut(&mut self) -> &mut HashMap<String, serde_json::Value> {
+        &mut self.additional_properties
+    }
+
+    /// Returns the predefined text data type.
```

Comment:
I'm not sure we want these functions here. I think we need a wider discussion about where these primitive Data Types live. Doesn't make sense to block this PR on that though so we can leave them.